i6_models/primitives/mixup.py

```diff
@@ -0,0 +1,153 @@
+            start_indicies_flat = torch.masked_select(
+                start_indicies, n_mask
+            )  # [B, M'] (M' denotes sum of num_mixup over the batch)
+
+            idx = torch.arange(t_dim)
+            idx = torch.unsqueeze(idx, dim=-1) + start_indicies_flat  # [T, M']
+            mixup_values = self.cache[idx]  # [T, M', F]
+            return mixup_values
+
+
+class Mixup:
+    """
+    Implement the Mixup data augmentation method
+    C.f. https://arxiv.org/abs/1710.09412
+
+    The code is partly based on Albert's implementation in returnn front-end
+    C.f. https://github.com/rwth-i6/i6_experiments/blob/main/users/zeyer/returnn/models/rf_mixup.py
+    """
+
+    def __init__(self, feature_dim: int, cfg: MixupConfig):
+        self.apply_prob = cfg.apply_prob
+        self.lambda_min = cfg.lambda_min
+        self.lambda_max = cfg.lambda_max
+        self.max_num_mix = cfg.max_num_mix
+        self.buffer_size = cfg.buffer_size
+        self.feature_dim = feature_dim
+
+        self.feature_buffer = FeatureBuffer(self.buffer_size, feature_dim)
+
+    def __call__(self, input: torch.Tensor, sequence_mask: torch.Tensor) -> torch.Tensor:
+        assert (
+            input.size()[-1] == self.feature_dim
+        ), "the given feature dimension does not match input feature dimension"
+        input_ = self._maybe_apply_mixup(input, sequence_mask)
+        self._append_to_buffer(input, sequence_mask)
+        return input_
+
+    def _append_to_buffer(self, input: torch.Tensor, sequence_mask: torch.Tensor):
+        _, _, f_dim = input.size()
+        # mask out the padded frames before append it to feature_buffer
+        input_flat = torch.masked_select(input, sequence_mask.unsqueeze(-1)).view(-1, f_dim)
+        self.feature_buffer.append(input_flat)
+
+    def _maybe_apply_mixup(self, input: torch.Tensor, sequence_mask: torch.Tensor) -> torch.Tensor:
+        if random.random() >= self.apply_prob:
+            return input
+
+        b_dim, t_dim, f_dim = input.size()  # [B, T, F]
+
+        max_end_idx = self.feature_buffer.buffer_size if self.feature_buffer.filled else self.feature_buffer.pos
+        if t_dim > max_end_idx:
+            return input
+
+        num_mixup = torch.randint(low=1, high=self.max_num_mix + 1, size=(b_dim,))  # [B]
+        max_num_mixup = max(num_mixup)  # [M] (M denotes maximum of num_mixup)
+
+        row_vector = torch.arange(0, max_num_mixup, 1)  # [M]
+        n_mask = torch.unsqueeze(num_mixup, dim=-1) > row_vector  # [B, M]
+
+        mixup_values = self.feature_buffer.getRandom(b_dim, t_dim, max_num_mixup, n_mask).to(
+            input.device
+        )  # [T, M', F] (M' denotes sum of num_mixup over the batch)
+
+        lambda_ = torch.FloatTensor(b_dim, max(num_mixup)).uniform_(self.lambda_min, self.lambda_max)  # [B, M]
+        mixup_scales = torch.FloatTensor(b_dim, max(num_mixup)).uniform_(0.001, 1)  # [B, M]
```

Comment:
This originates from Zoltans implementation. I'm not sure what would be the best. I think the scales themselves make sense on a linear scale (when we tune other scales, we also always do grid search on linear scale).

Reply:
I don't quite understand the lambda/mixup_scales. On a high level we mix the original audio with some sequence-noise which is some other audio in the corpus. The mixing should be done with different signal to noise ratios (SNR). And I would select the SNR on a log-uniform scale between some reasonable minimum (barely audible) and maximum (almost 1).

Is the lambda the selector for SNR and the mixup_scales is the weight how the 1,...,n individual noise sequences are mixed to get "the noise"? Then I would agree this should be uniform. And we can even do uniform from [0,1] - no need for lower cutoff.

Reply:
First think about the simple case, you take only one slice from the mixup cache. This is added to the input via same scale `lambda_` taken randomly from Uniform([lambda_min,lambda_max]), like `input = input + lambda_ * mixup_values`.

To have more noise, we don't just take a single slice, but multiple. But no matter how much are taken (as this is also a random var), we want that the total magnitude of what we add stays the same, i.e. like `lambda_`. So if we take N slices, we could just sample N `lambda_`s, or say `lambda_` with shape [N]. Now we could just weight the `lambda_` with 1/N each, such that when we sum over them, the total weight is still in [lambda_min,lambda_max].

Now, to add a bit more randomness, instead of using weights 1/N, we have these `mixup_scales`, and then multiply the `lambda_`s by `mixup_scales/sum(mixup_scales)`. I.e. again this makes sure that the total weight is still in [lambda_min,lambda_max], but just with even more randomness.